### DISCUSSION TITLE
JWT bearer token limit?

### DISCUSSION THREAD

**Discussion by ArKam** [dani-garcia/vaultwarden#3705] [Q&A] [answered]
Hi everyone,

I’m currently working on a vaultwarden module for ansible, but I’m kinda having a bit of troubles checking my head around the way the vault works.

Currently, I’m able to authenticate against the vault server with a user’s personal ID and secret (from the API key page) and retrieve a bearer token from the API’s identity endpoint itself.

With this token I’m able to request the server for either a cipher, a collection etc.

My main issue is to know if I can decode/decrypt the resulting encrypted payload just using the bearer token or is my user required to compute the decryption key from its master password?

If so, what’s the point of authenticating using the bearer token at all as all request?

Did I missed something here?

Because currently the automated process seems a bit odd, as my module would need the user to log on a client to retrieve the API key, then use the API key within my ansible module to retrieve the bearer token, that is then used to retrieve the cipher or the vault through the API’s sync endpoint and then again to either provide the module with the master password to decrypt ciphers and I’d like to avoid the module to request for the master password if possible.

**Comment by BlackDex** [comment, marked as answer]
You still need the master password to be able to decrypt everything. It only provides access to the vault items, not to decrypt them.
The reason for this is, that if you use the master-password and email address to login, and that traffic is intercepted in some way, that data can be used to decrypt the vault also.

If you use the API-Key to authenticate, and locally after the retrieval of all the vault data (or a specific vault item) you can then decrypt it using the password, without the password ever leaving the client even thought it would be hashed.

So, yes, you would still need the plain/text password to be able to decrypt. But it does provide some security into the call to the backend not having the password as a variable (again the password is never sent in plain text)

**Comment by ArKam** [comment]
Ok ok, I did get the process correctly tho.
Last question, why does the identity api send public and private key with the bearer token? What are those keys made for?

**Reply by BlackDex** [reply]
That is used for accessing the Organization vault items if I'm correct.

I suggest to read https://bitwarden.com/images/resources/security-white-paper-download.pdf for a more detailed explanation.

**Reply by ArKam** [reply]
Thanks a lot, I’ll read that as It may help find a way to avoid having to ask the user for two set of credentials and password.
